python/paddle/fluid/framework.py

```diff
@@ -3192,7 +3192,13 @@ def network(is_test):
         The two code snippets above will generate and print same programs.
         """
         if for_test:
-            p = self._inference_optimize(prune_read_op=False)
+            if self._appending_grad_times > 0:
```

Comment:
in order to check whether the program contains backward part, the flag `_appending_grad_times` can be used to judge here, but it also can judge by other way, like

```
def _has_backward_op(self):
    op_role_key = core.op_proto_and_checker_maker.kOpRoleAttrName()
    backward = int(core.op_proto_and_checker_maker.OpRole.Backward)
    for op in reversed(self.global_block().ops):
        if int(op.all_attrs()[op_role_key]) == backward
            return true
    return false
```

which is better?

Reply:
I am not sure, maybe both ok here.